src/Language/PureScript/TypeChecker.hs

```diff
@@ -152,18 +153,30 @@ checkDuplicateTypeArguments args = for_ firstDup $ \dup ->
   firstDup = listToMaybe $ args \\ nub args
 
 checkTypeClassInstance
-  :: (MonadState CheckState m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
-  => ModuleName
+  :: (MonadState CheckState m, MonadError MultipleErrors m)
+  => TypeClassData
+  -> Int -- type argument index
   -> Type
   -> m ()
-checkTypeClassInstance _ (TypeVar _) = return ()
-checkTypeClassInstance _ (TypeLevelString _) = return ()
-checkTypeClassInstance _ (TypeConstructor ctor) = do
-  env <- getEnv
-  when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
-  return ()
-checkTypeClassInstance m (TypeApp t1 t2) = checkTypeClassInstance m t1 >> checkTypeClassInstance m t2
-checkTypeClassInstance _ ty = throwError . errorMessage $ InvalidInstanceHead ty
+checkTypeClassInstance cls i = check where
+  -- check if the argument is determined by a single fundep group
+  isDeterminedInGroup :: FunctionalDependency -> Bool
+  isDeterminedInGroup fd = not (i `elem` fdDeterminers fd) && (i `elem` fdDetermined fd)
+
+  -- is this argument fully determined via fundeps
+  isFunDepDetermined = Just (All True) == foldMap (Just . All . isDeterminedInGroup) (typeClassDependencies cls)
```

Comment:
Or is that the transitive case you're talking about?  Although isn't *"just by some type variable which is not itself determined"* inherently transitive?